tests/test_url_shortening.py

```diff
@@ -0,0 +1,206 @@
+
+    def setUp(self):
+        """Setup app for testing before each test case."""
+        self.app = create_app('testing')
+        self.app_context = self.app.app_context()
+        self.app_context.push()
+        db.create_all()
+        self.client = self.app.test_client(use_cookies=True)
+        self.user = User(first_name='ichiato', last_name='ikikin',
+                         email='ichiato@yahoo.com', password='password')
+        self.post_data = json.dumps({
+            "url": "http://www.andela.com",
+            })
+        self.post_data2 = json.dumps({})
+        self.post_data3 = json.dumps({
+            "url": "http://www.andela.com",
+            "vanity_string": "python"
+            })
+
+    def tearDown(self):
+        """Delete app and db instances after each test case."""
+        db.session.remove()
+        db.drop_all()
+        self.app_context.pop()
+
+    def api_email_auth_headers(self, email, password):
+        """Setup an authorization header with an email and password."""
+        return {
+            'Authorization': 'Basic ' + b64encode(
+                (email + ':' + password).encode('utf-8')).decode('utf-8'),
+            'Accept': 'application/json',
+            'Content-Type': 'application/json'
+        }
+
+    def api_token_auth_headers(self, token):
+        """Setup an authorization header with a token."""
+        return {
+            'Authorization': 'Basic ' + b64encode(
+                (token + ':' + '').encode('utf-8')).decode('utf-8'),
+            'Accept': 'application/json',
+            'Content-Type': 'application/json'
+        }
+
+    def test_anonymous_user_shorten_via_api(self):
+        """Test that anonymous users cannot shorten urls via the API."""
+        response = self.client.post(url_for('api.shorten_url'),
+                                    data=self.post_data)
+        json_response = json.loads(response.data.decode('utf-8'))
+        msg = json_response['message']
+        self.assertEqual(msg, 'Permission required! You are not allowed'
+                         ' access to this resource.')
+        self.assertEqual(403, response.status_code)
+
+    def test_registered_user_shorten_via_api(self):
+        """Test that registered users can shorten urls via the API."""
+        self.user.save()
+        email_header = self.api_email_auth_headers('ichiato@yahoo.com',
+                                                   'password')
+        token_response = self.client.get(url_for('api.get_token'),
+                                         headers=email_header)
+        json_token_response = json.loads(token_response.data.decode('utf-8'))
+        token = json_token_response['token']
+        token_header = self.api_token_auth_headers(token)
+        shorten_response_with_email_auth = self.client.post(
+            url_for('api.shorten_url'), headers=email_header,
+            data=self.post_data)
+        shorten_response_with_token_auth = self.client.post(
```

Comment:
Since this is the actual endpoint you want to test, focus on it and avoid calling the other ones above. You could still obtain the token by calling the function that generates it directly as you did in the user model test.